File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/array/matrix/P151_ReverseWordsInAString.java
package com.dsa.leetcode.labuladong.array.matrix;

public class P151_ReverseWordsInAString {
    public static void main(String[] args) {
        P151_ReverseWordsInAString p151 = new P151_ReverseWordsInAString();
        String s = p151.reverseWords("  hello world  ");
        System.out.println(s);
    }

    public String reverseWords(String s) {
        StringBuilder sb = trimSpace(s);
        revese(0, sb.length() - 1, sb);
        reverseEachWord(sb);
        return sb.toString();
    }

    private void reverseEachWord(StringBuilder sb) {
        int left = 0;
        int right = 0;
        int n = sb.length();
        while (right < n) {
            while (right < n && sb.charAt(right) != ' ') {
                right++;
            }
            revese(left, right - 1, sb);
            left = ++right;
        }
    }

    private void revese(int left, int right, StringBuilder sb) {
        char temp;
        while (left < right) {
            temp = sb.charAt(left);
            sb.setCharAt(left, sb.charAt(right));
            sb.setCharAt(right, temp);
            left++;
            right--;
        }
    }

    private StringBuilder trimSpace(String s) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0, n = s.length(); i < n; i++) {
            if (s.charAt(i) == ' ') {
                continue;
            }
            while (i < n && s.charAt(i) != ' ') {
                sb.append(s.charAt(i));
                i++;
            }
            sb.append(' ');
        }
        return sb.deleteCharAt(sb.length() - 1);
    }

    public String reverseWords2(String s) {
        char[] chars = s.toCharArray();
        int n = chars.length;
        int right;

        char[] newChars = new char[n + 1];
        int index = 0;

        for (int left = n - 1; left >= 0; left--) {
            if (chars[left] == ' ') {
                continue;
            }
            // 不是空字符串
            right = left;
            while (left >= 0 && chars[left] != ' ') {
                left--;
            }
            // 复制字符串，注意这里的left + 1
            for (int i = left + 1; i <= right; i++) {
                newChars[index++] = chars[i];
            }
            newChars[index++] = ' ';
        }
        return String.valueOf(newChars, 0, index - 1);
    }
}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/array/matrix/P151_ReverseWordsInAString.py
class Solution:
    def reverseWords(self, s: str) -> str:
        words = self.trimSpaces(s)
        self.reverse(0, len(words) - 1, words)
        self.reverseEachWord(words)
        return "".join(words)

    def trimSpaces(self, s: str):
        words = []
        n, i = len(s), 0
        while i < n:
            if s[i] != " ":
                while i < n and s[i] != " ":
                    words.append(s[i])
                    i += 1
                words.append(" ")
            i += 1
        return words[0:len(words) - 1]

    def reverse(self, left, right, words):
        while left < right:
            temp = words[left]
            words[left] = words[right]
            words[right] = temp
            left += 1
            right -= 1

    def reverseEachWord(self, words):
        n = len(words)
        left, right = 0, 0
        while right < n:
            while right < n and words[right] != " ":
                right += 1
            self.reverse(left, right - 1, words)
            right += 1
            left = right


if __name__ == '__main__':
    s = "  hello world  "
    solu = Solution()
    words = solu.reverseWords(s)
    print(words)
